sensor/kubernetes/eventpipeline/component/component.go

```diff
@@ -12,15 +12,27 @@ type PipelineComponent interface {
 	Stop(error)
 }
 
-// Resolver component that performs the dependency resolution
+// Resolver component that performs the dependency resolution. This is the core component of the pipeline, and the reason
+// why re-sync is no longer required in kubernetes informers. It receives inputs from any kubernetes handler (from
+// listeners in sensor/kubernetes/listener/resources) in the form of a ResourceEvent object. Its main responsibility
+// is to look at the DeploymentReference slice, and process any deployments that need to be updated. Deployment references
+// are based on resolver.DeploymentResolution callback functions, which are defined by the handlers.
+//
+// For example: given deployment-A which has a service account (sa-A), and RBAC resources attached to it. When any of the
+// RBAC resources bound to sa-A change (created, updated or deleted), deployment-A has to be reprocessed. This is required
+// because the storage.Deployment object contains properties that are based off RBAC resources. They also need reprocessing,
+// because the state of violations might change. Rather than reprocessing every deployment every minute, the Resolver component
+// will look at any DeploymentReference entries that handlers might have added, and enqueue those deployment for reprocessing.
 //
 //go:generate mockgen-wrapper
 type Resolver interface {
 	PipelineComponent
 	Send(event *ResourceEvent)
 }
 
-// OutputQueue component that redirects Resource Events and Alerts to the output channel
+// OutputQueue component that redirects Resource Events and Alerts to the output channel. This component is the last step
```

Comment:
What is the output channel responsible for? Is it a channel which sends the transformed, processed and enriched Kubernetes objects as protos to other components like Sensor, Detector and so on?
Generally speaking we never operate directly on the Kubernetes resources but always need pre-processing before anything is done?

Reply:
> What is the output channel responsible for?

I assume you mean the OutputQueue component. This isn't _just_ a go channel.

>  Is it a channel which sends the transformed, processed and enriched Kubernetes objects as protos to other components like Sensor, Detector and so on?

Yes.

To be more specific, the output component has a channel inside (`forwardQueue`) and its messages are simply forwarded to the pipeline `ReponsesC` channel. Every component in sensor can implement a `.ResponseC` function that returns a channel with gRPC messages. So the output component is responsible for making these messages that will be eventually picked up by sensor gRPC component. 

Interacting with the dector is a bit different, it uses a "push" model. So the output component will also send deployment directly to the detector if they are marked for detection. The code that does that in `output_impl.go` is pretty straightforward:

```go
for _, detectorRequest := range msg.DetectorMessages {
        q.detector.ProcessDeployment(msg.Context, detectorRequest.Object, detectorRequest.Action)
}
```

Reply:
Thx for the clarification, this makes sense to me now!